### PR TITLE
[SPARK-35832][CORE][ML][K8S][TESTS] Add LocalRootDirsTest trait

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

To make the test suite more robust, this PR aims to add a new trait, `LocalRootDirsTest`, by refactoring `SortShuffleSuite`'s helper functions and applying it to the following:
- ShuffleNettySuite
- ShuffleOldFetchProtocolSuite
- ExternalShuffleServiceSuite
- KubernetesLocalDiskShuffleDataIOSuite
- LocalDirsSuite
- RDDCleanerSuite
- ALSCleanerSuite

In addition, this fixes a UT in `KubernetesLocalDiskShuffleDataIOSuite`.

### Why are the changes needed?

`ShuffleSuite` is extended by four classes but only `SortShuffleSuite` does the clean-up correctly.
```
ShuffleSuite
- SortShuffleSuite
- ShuffleNettySuite
- ShuffleOldFetchProtocolSuite
- ExternalShuffleServiceSuite
```

Since `KubernetesLocalDiskShuffleDataIOSuite` is looking for the other storage directory, the leftover of `ShuffleSuite` causes flakiness.
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-3.2/2649/testReport/junit/org.apache.spark.shuffle/KubernetesLocalDiskShuffleDataIOSuite/recompute_is_not_blocked_by_the_recovery/
```
org.apache.spark.SparkException: Job aborted due to stage failure: task 0.0 in stage 1.0 (TID 3) had a not serializable result: org.apache.spark.ShuffleSuite$NonJavaSerializableClass
...
org.apache.spark.shuffle.KubernetesLocalDiskShuffleDataIOSuite.$anonfun$new$2(KubernetesLocalDiskShuffleDataIOSuite.scala:52)
```

For the other suites, the clean-up implementation is used but not complete. So, they are refactored to use new trait.

### Does this PR introduce _any_ user-facing change?

No, this is a test-only change.

### How was this patch tested?

Pass the CIs.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this when you have some time, @HyukjinKwon and @viirya ?

File: core/src/test/scala/org/apache/spark/LocalRootDirsTest.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+package org.apache.spark
+
+import java.io.File
+import java.util.UUID
+
+import org.apache.spark.util.Utils
+
+
+trait LocalRootDirsTest extends SparkFunSuite with LocalSparkContext {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/LocalRootDirsTest.scala:26]: This functionality came from `SortShuffleSuite`.

**Comment by dongjoon-hyun** [comment]
This is also a patch to fix flaky tests, @gengliangwang .

File: core/src/test/scala/org/apache/spark/LocalRootDirsTest.scala (hunk 1 above)
Comment (viirya) [reply core/src/test/scala/org/apache/spark/LocalRootDirsTest.scala:26]: Can we add some comments for this trait? So others in the future can know quickly what this trait is for.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleDataIOSuite.scala [hunk 2]
```diff
@@ -192,7 +195,7 @@ class KubernetesLocalDiskShuffleDataIOSuite extends SparkFunSuite with LocalSpar
       val rdd3 = rdd2.reduceByKey(_ + _)
       val rdd4 = rdd3.sortByKey()
 
-      assert(rdd4.count() === 3)
+      assert(rdd4.count() === 2)
```
Comment (viirya) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleDataIOSuite.scala:198]: why this count changes?

**Review by viirya — COMMENTED** [review]
Looks good. I have a question about a change looks unrelated to the cleaning up.

**Comment by dongjoon-hyun** [comment]
Thank you, @viirya . Yes, that was a hidden bug and uncovered by this. I mentioned in the PR description like the following.
> In addition, this fixes a UT in KubernetesLocalDiskShuffleDataIOSuite.

Merged to master.

**Comment by viirya** [comment]
Thank you @dongjoon-hyun

**Comment by HyukjinKwon** [comment]
lgtm2

**Pull request closed without merging**